**BMW Car Sales Data Analysis**

🌟 **Task 2: Data Cleaning and Pre-processing**

**Performing data cleaning process by handling missing data, duplicates, formatting, derived features, and transformation.**

**Step 1:  Data Loading and Initial Overview**

● Importing the dataset using Pandas and providing an overview:

○ Number of rows and columns

○ Data types of each column

○ Initial observations (e.g., head(), info(), describe())

**Step 2: Data Pre-processing**

● Performing all necessary cleaning steps :

○ Handling missing values

○ Removing duplicates

○ Correcting data types

○ Creating derived columns

○ Filtering or aggregating data


**Importing Libraries**

In [1]:
import numpy as np

In [2]:
import pandas as pd

**Step 1- Data Loading and Initial Overview**

**Reading the dataset**

In [3]:
df=pd.read_csv("BMW Car.csv")

In [4]:
df

Model  Year         Region   Color Fuel_Type Transmission  \
0      5 Series  2016           Asia     Red    Petrol       Manual   
1            i8  2013  North America     Red    Hybrid    Automatic   
2      5 Series  2022  North America    Blue    Petrol    Automatic   
3            X3  2024    Middle East    Blue    Petrol    Automatic   
4      7 Series  2020  South America   Black    Diesel       Manual   
...         ...   ...            ...     ...       ...          ...   
49995        i3  2014           Asia     Red    Hybrid       Manual   
49996        i3  2023    Middle East  Silver  Electric       Manual   
49997  5 Series  2010    Middle East     Red    Petrol    Automatic   
49998        i3  2020           Asia   White  Electric    Automatic   
49999        X1  2020  North America    Blue    Diesel       Manual   

       Engine_Size_L  Mileage_KM  Price_USD  Sales_Volume Sales_Classification  
0                3.5      151748      98740          8300                 High  
1                1.6      121671      79219          3428                  Low  
2                4.5       10991     113265          6994                  Low  
3                1.7       27255      60971          4047                  Low  
4                2.1      122131      49898          3080                  Low  
...              ...         ...        ...           ...                  ...  
49995            4.6      151030      42932          8182                 High  
49996            4.2      147396      48714          9816                 High  
49997            4.5      174939      46126          8280                 High  
49998            3.8        3379      58566          9486                 High  
49999            3.3      171003      77492          1764                  Low  

[50000 rows x 11 columns]

**Determining Number of rows and columns count**

In [5]:
print("Number of rows=",df.shape[0])

Number of rows= 50000


In [6]:
print("Number of rows=",len(df.columns))

Number of rows= 11


In [7]:
df.shape

(50000, 11)

**Getting the Data types of each column**

In [8]:
df.dtypes

Model                    object
Year                      int64
Region                   object
Color                    object
Fuel_Type                object
Transmission             object
Engine_Size_L           float64
Mileage_KM                int64
Price_USD                 int64
Sales_Volume              int64
Sales_Classification     object
dtype: object

**Initial observations of datatype**

**Loading First five rows for getting the overall idea of dataset**

In [9]:
df.head()

Model  Year         Region  Color Fuel_Type Transmission  Engine_Size_L  \
0  5 Series  2016           Asia    Red    Petrol       Manual            3.5   
1        i8  2013  North America    Red    Hybrid    Automatic            1.6   
2  5 Series  2022  North America   Blue    Petrol    Automatic            4.5   
3        X3  2024    Middle East   Blue    Petrol    Automatic            1.7   
4  7 Series  2020  South America  Black    Diesel       Manual            2.1   

   Mileage_KM  Price_USD  Sales_Volume Sales_Classification  
0      151748      98740          8300                 High  
1      121671      79219          3428                  Low  
2       10991     113265          6994                  Low  
3       27255      60971          4047                  Low  
4      122131      49898          3080                  Low

**Formatting the code**

In [10]:
## Strip and standardize column values

df['Model'] = df['Model'].str.title().str.strip()
df['Region'] = df['Region'].str.title().str.strip()
df['Color'] = df['Color'].str.title().str.strip()
df['Fuel_Type'] = df['Fuel_Type'].str.title().str.strip()
df['Transmission'] = df['Transmission'].str.title().str.strip()
df['Sales_Classification'] = df['Sales_Classification'].str.title().str.strip()

## Round and clean numeric columns

df['Engine_Size_L'] = df['Engine_Size_L'].round(1)
df['Price_USD'] = df['Price_USD'].round(2)

In [11]:
df

Model  Year         Region   Color Fuel_Type Transmission  \
0      5 Series  2016           Asia     Red    Petrol       Manual   
1            I8  2013  North America     Red    Hybrid    Automatic   
2      5 Series  2022  North America    Blue    Petrol    Automatic   
3            X3  2024    Middle East    Blue    Petrol    Automatic   
4      7 Series  2020  South America   Black    Diesel       Manual   
...         ...   ...            ...     ...       ...          ...   
49995        I3  2014           Asia     Red    Hybrid       Manual   
49996        I3  2023    Middle East  Silver  Electric       Manual   
49997  5 Series  2010    Middle East     Red    Petrol    Automatic   
49998        I3  2020           Asia   White  Electric    Automatic   
49999        X1  2020  North America    Blue    Diesel       Manual   

       Engine_Size_L  Mileage_KM  Price_USD  Sales_Volume Sales_Classification  
0                3.5      151748      98740          8300                 High  
1                1.6      121671      79219          3428                  Low  
2                4.5       10991     113265          6994                  Low  
3                1.7       27255      60971          4047                  Low  
4                2.1      122131      49898          3080                  Low  
...              ...         ...        ...           ...                  ...  
49995            4.6      151030      42932          8182                 High  
49996            4.2      147396      48714          9816                 High  
49997            4.5      174939      46126          8280                 High  
49998            3.8        3379      58566          9486                 High  
49999            3.3      171003      77492          1764                  Low  

[50000 rows x 11 columns]

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Model                 50000 non-null  object 
 1   Year                  50000 non-null  int64  
 2   Region                50000 non-null  object 
 3   Color                 50000 non-null  object 
 4   Fuel_Type             50000 non-null  object 
 5   Transmission          50000 non-null  object 
 6   Engine_Size_L         50000 non-null  float64
 7   Mileage_KM            50000 non-null  int64  
 8   Price_USD             50000 non-null  int64  
 9   Sales_Volume          50000 non-null  int64  
 10  Sales_Classification  50000 non-null  object 
dtypes: float64(1), int64(4), object(6)
memory usage: 4.2+ MB


In [13]:
df.describe()

Year  Engine_Size_L     Mileage_KM      Price_USD  Sales_Volume
count  50000.000000   50000.000000   50000.000000   50000.000000  50000.000000
mean    2017.015700       3.247180  100307.203140   75034.600900   5067.514680
std        4.324459       1.009078   57941.509344   25998.248882   2856.767125
min     2010.000000       1.500000       3.000000   30000.000000    100.000000
25%     2013.000000       2.400000   50178.000000   52434.750000   2588.000000
50%     2017.000000       3.200000  100388.500000   75011.500000   5087.000000
75%     2021.000000       4.100000  150630.250000   97628.250000   7537.250000
max     2024.000000       5.000000  199996.000000  119998.000000   9999.000000

**Step 2- Data Pre-processing**

**Handling missing vslues**

In [14]:
df.isnull()

Model   Year  Region  Color  Fuel_Type  Transmission  Engine_Size_L  \
0      False  False   False  False      False         False          False   
1      False  False   False  False      False         False          False   
2      False  False   False  False      False         False          False   
3      False  False   False  False      False         False          False   
4      False  False   False  False      False         False          False   
...      ...    ...     ...    ...        ...           ...            ...   
49995  False  False   False  False      False         False          False   
49996  False  False   False  False      False         False          False   
49997  False  False   False  False      False         False          False   
49998  False  False   False  False      False         False          False   
49999  False  False   False  False      False         False          False   

       Mileage_KM  Price_USD  Sales_Volume  Sales_Classification  
0           False      False         False                 False  
1           False      False         False                 False  
2           False      False         False                 False  
3           False      False         False                 False  
4           False      False         False                 False  
...           ...        ...           ...                   ...  
49995       False      False         False                 False  
49996       False      False         False                 False  
49997       False      False         False                 False  
49998       False      False         False                 False  
49999       False      False         False                 False  

[50000 rows x 11 columns]

In [15]:
df.isnull().sum()

Model                   0
Year                    0
Region                  0
Color                   0
Fuel_Type               0
Transmission            0
Engine_Size_L           0
Mileage_KM              0
Price_USD               0
Sales_Volume            0
Sales_Classification    0
dtype: int64

**Dropping the duplicate values**

In [16]:
df.drop_duplicates(inplace=True)

No Missing Values, Duplicates Found. No issue with Datatypes

**Derived Features**

**Creating derived columns**

In [17]:
df['Car_Age'] = 2025 - df['Year']

In [18]:
df['Price_per_KM'] = df['Price_USD'] / df['Mileage_KM']

In [19]:
df

Model  Year         Region   Color Fuel_Type Transmission  \
0      5 Series  2016           Asia     Red    Petrol       Manual   
1            I8  2013  North America     Red    Hybrid    Automatic   
2      5 Series  2022  North America    Blue    Petrol    Automatic   
3            X3  2024    Middle East    Blue    Petrol    Automatic   
4      7 Series  2020  South America   Black    Diesel       Manual   
...         ...   ...            ...     ...       ...          ...   
49995        I3  2014           Asia     Red    Hybrid       Manual   
49996        I3  2023    Middle East  Silver  Electric       Manual   
49997  5 Series  2010    Middle East     Red    Petrol    Automatic   
49998        I3  2020           Asia   White  Electric    Automatic   
49999        X1  2020  North America    Blue    Diesel       Manual   

       Engine_Size_L  Mileage_KM  Price_USD  Sales_Volume  \
0                3.5      151748      98740          8300   
1                1.6      121671      79219          3428   
2                4.5       10991     113265          6994   
3                1.7       27255      60971          4047   
4                2.1      122131      49898          3080   
...              ...         ...        ...           ...   
49995            4.6      151030      42932          8182   
49996            4.2      147396      48714          9816   
49997            4.5      174939      46126          8280   
49998            3.8        3379      58566          9486   
49999            3.3      171003      77492          1764   

      Sales_Classification  Car_Age  Price_per_KM  
0                     High        9      0.650684  
1                      Low       12      0.651092  
2                      Low        3     10.305250  
3                      Low        1      2.237057  
4                      Low        5      0.408561  
...                    ...      ...           ...  
49995                 High       11      0.284261  
49996                 High        2      0.330497  
49997                 High       15      0.263669  
49998                 High        5     17.332347  
49999                  Low        5      0.453162  

[50000 rows x 13 columns]

**Filtering and aggregating data**

**1. By Year**

In [20]:
# Cars manufactured after 2020
df[df['Year'] >= 2020]

Model  Year         Region   Color Fuel_Type Transmission  \
2      5 Series  2022  North America    Blue    Petrol    Automatic   
3            X3  2024    Middle East    Blue    Petrol    Automatic   
4      7 Series  2020  South America   Black    Diesel       Manual   
6            I8  2022         Europe   White    Diesel       Manual   
12     7 Series  2020  North America  Silver    Diesel    Automatic   
...         ...   ...            ...     ...       ...          ...   
49980        I3  2022           Asia   White    Hybrid    Automatic   
49982        M3  2024           Asia  Silver    Hybrid    Automatic   
49996        I3  2023    Middle East  Silver  Electric       Manual   
49998        I3  2020           Asia   White  Electric    Automatic   
49999        X1  2020  North America    Blue    Diesel       Manual   

       Engine_Size_L  Mileage_KM  Price_USD  Sales_Volume  \
2                4.5       10991     113265          6994   
3                1.7       27255      60971          4047   
4                2.1      122131      49898          3080   
6                1.8      196741      55064          7949   
12               3.8       27403     100015          8111   
...              ...         ...        ...           ...   
49980            5.0       38575     112981          6123   
49982            1.7      198846      85034          7888   
49996            4.2      147396      48714          9816   
49998            3.8        3379      58566          9486   
49999            3.3      171003      77492          1764   

      Sales_Classification  Car_Age  Price_per_KM  
2                      Low        3     10.305250  
3                      Low        1      2.237057  
4                      Low        5      0.408561  
6                     High        3      0.279881  
12                    High        5      3.649783  
...                    ...      ...           ...  
49980                  Low        3      2.928866  
49982                 High        1      0.427637  
49996                 High        2      0.330497  
49998                 High        5     17.332347  
49999                  Low        5      0.453162  

[16701 rows x 13 columns]

In [21]:
# Cars manufactured before 2020
df[df['Year'] < 2020]

Model  Year         Region   Color Fuel_Type Transmission  \
0      5 Series  2016           Asia     Red    Petrol       Manual   
1            I8  2013  North America     Red    Hybrid    Automatic   
5      5 Series  2017    Middle East  Silver    Diesel       Manual   
7            M5  2014           Asia   Black    Diesel    Automatic   
8            X3  2016  South America   White    Diesel    Automatic   
...         ...   ...            ...     ...       ...          ...   
49992        I3  2017    Middle East    Blue    Petrol    Automatic   
49993  5 Series  2013         Europe    Grey    Petrol       Manual   
49994  5 Series  2010  North America  Silver    Petrol       Manual   
49995        I3  2014           Asia     Red    Hybrid       Manual   
49997  5 Series  2010    Middle East     Red    Petrol    Automatic   

       Engine_Size_L  Mileage_KM  Price_USD  Sales_Volume  \
0                3.5      151748      98740          8300   
1                1.6      121671      79219          3428   
5                1.9      171362      42926          1232   
7                1.6      121156     102778           632   
8                1.7       48073     116482          8944   
...              ...         ...        ...           ...   
49992            3.9       79874     115046          9015   
49993            4.7      155013      80263          3700   
49994            3.6       56672      32990          1575   
49995            4.6      151030      42932          8182   
49997            4.5      174939      46126          8280   

      Sales_Classification  Car_Age  Price_per_KM  
0                     High        9      0.650684  
1                      Low       12      0.651092  
5                      Low        8      0.250499  
7                      Low       11      0.848311  
8                     High        9      2.423023  
...                    ...      ...           ...  
49992                 High        8      1.440344  
49993                  Low       12      0.517782  
49994                  Low       15      0.582122  
49995                 High       11      0.284261  
49997                 High       15      0.263669  

[33299 rows x 13 columns]

**Insight: More number of cars sold out before the year 2020 while comparing with the one on and after 2020**

In [22]:
for year in range(2010, 2025):
    count = len(df[df['Year'] == year])
    print(f"Cars produced in {year}: {count}")

Cars produced in 2010: 3330
Cars produced in 2011: 3278
Cars produced in 2012: 3332
Cars produced in 2013: 3326
Cars produced in 2014: 3350
Cars produced in 2015: 3358
Cars produced in 2016: 3365
Cars produced in 2017: 3290
Cars produced in 2018: 3278
Cars produced in 2019: 3392
Cars produced in 2020: 3205
Cars produced in 2021: 3372
Cars produced in 2022: 3478
Cars produced in 2023: 3219
Cars produced in 2024: 3427


**Insight: Highest number of cars produced in : 2022
Count: 3478**

🔍 Key Insights:
🔹 1. Consistent Production With Slight Growth
BMW maintained a very consistent production volume of ~3,200 to ~3,400 cars annually.

Suggests a stable supply chain and market demand, without major fluctuations.

🔹 2. 2020 Production Dip
2020 shows the lowest output (3,205), likely due to:

COVID-19 pandemic impact

Global disruptions in manufacturing, logistics, and demand

🔹 3. Post-Pandemic Recovery
Production increased again from 2021, peaking in 2022 (3,478 cars) — the highest in the dataset.

Indicates strong recovery and demand rebound post-pandemic.

🔹 4. Slight Dip in 2023, Rebound in 2024
2023 shows a mild drop (3,219), followed by a strong return in 2024 (3,427).

Reflects dynamic adaptation to market cycles or strategy shifts.

🧭 Strategic Takeaways:
✅ BMW has shown resilience and adaptability in maintaining production levels through global challenges.

🔄 Monitoring this trend helps in forecasting demand and planning inventory.


**By Fuel Type**

In [24]:
count_diesel = len(df[df['Fuel_Type'] == 'Diesel'])
print("Number of Diesel cars:", count_diesel)

Number of Diesel cars: 12263


In [25]:
count_petrol = len(df[df['Fuel_Type'] == 'Petrol'])
print("Number of Petrol  cars:", count_petrol)


Number of Petrol  cars: 12550


In [27]:
count_Hybrid = len(df[df['Fuel_Type'] == 'Hybrid'])
print("Number of Hybrid cars:", count_Hybrid)

Number of Hybrid cars: 12716


In [42]:
count_Electric = len(df[df['Fuel_Type'] == 'Electric'])
print("Number of electric cars:", count_Electric)

Number of electric cars: 12471


In [55]:
df[df['Year'] < 2020].groupby('Fuel_Type')[['Sales_Volume', 'Price_USD']].mean().reset_index()


Fuel_Type  Sales_Volume     Price_USD
0    Diesel   5081.506444  75206.454646
1  Electric   5047.534632  75176.756133
2    Hybrid   5066.788191  74850.611052
3    Petrol   5041.027191  74726.664758

In [53]:
df[df['Year'] >= 2020].groupby('Fuel_Type')[['Sales_Volume', 'Price_USD']].mean().reset_index()


Fuel_Type  Sales_Volume     Price_USD
0    Diesel   5092.999271  74829.135083
1  Electric   5098.042599  75475.571119
2    Hybrid   5090.895662  74692.415944
3    Petrol   5055.255942  75521.412965

**Insight:

Fuel Type in high demand over 2010-2024:  Hybrid

Fuel Type in high demand over before 2020 : Diesel

Fuel Type in high demand after 2020: Electric**

**1. 2010–2024: Hybrid Takes the Lead**

Over the full period, Hybrid vehicles dominate.

Reflects the transition phase between traditional engines and full electrification.

Hybrid cars combine fuel efficiency with familiar driving range and performance, making them appealing across markets.

🔹 **2. Before 2020: Diesel Was King**

Diesel led in early years due to:

Better mileage

Strong torque for performance and towing (especially in SUVs)

However, regulatory pressures and emission scandals (e.g., Dieselgate) caused demand to decline approaching 2020.

🔹 **3. Post-2020: Rise of Electric**

Electric cars gained momentum after 2020, likely driven by:

Stricter emissions regulations

Government incentives and EV infrastructure improvements

Changing consumer sentiment toward sustainability

Models like the i3 and i8 likely contributed to this rise.


Understanding what models and specs are in demand.

Filter: Cars with Sales_Classification = "High"

Aggregate: Count of such cars by Model and Fuel_Type

In [48]:
# High selling models by fuel type
df[df['Sales_Classification'] == 'High'].groupby(['Model','Fuel_Type']).size().reset_index(name='Count')


Model Fuel_Type  Count
0   3 Series    Diesel   1134
1   3 Series  Electric   1159
2   3 Series    Hybrid   1203
3   3 Series    Petrol   1099
4   5 Series    Diesel   1128
5   5 Series  Electric   1136
6   5 Series    Hybrid   1143
7   5 Series    Petrol   1185
8   7 Series    Diesel   1169
9   7 Series  Electric   1131
10  7 Series    Hybrid   1206
11  7 Series    Petrol   1160
12        I3    Diesel   1202
13        I3  Electric   1138
14        I3    Hybrid   1130
15        I3    Petrol   1148
16        I8    Diesel   1114
17        I8  Electric   1152
18        I8    Hybrid   1172
19        I8    Petrol   1168
20        M3    Diesel   1069
21        M3  Electric   1098
22        M3    Hybrid   1154
23        M3    Petrol   1092
24        M5    Diesel   1044
25        M5  Electric   1131
26        M5    Hybrid   1175
27        M5    Petrol   1128
28        X1    Diesel   1123
29        X1  Electric   1183
30        X1    Hybrid   1141
31        X1    Petrol   1123
32        X3    Diesel   1102
33        X3  Electric   1121
34        X3    Hybrid   1150
35        X3    Petrol   1124
36        X5    Diesel   1112
37        X5  Electric   1110
38        X5    Hybrid   1098
39        X5    Petrol   1167
40        X6    Diesel   1066
41        X6  Electric   1112
42        X6    Hybrid   1144
43        X6    Petrol   1156

✅ Most demanded model-fuel combination:
BMW 7 Series – Hybrid with 1,206 units sold

🔍 Detailed Insights:

🔸 1. 7 Series Dominates in Hybrid Demand

The 7 Series Hybrid tops all model-fuel combinations.

Indicates strong market preference for luxury + hybrid efficiency.

Reflects shifting trends: consumers want prestige + eco-friendliness.

🔸 2. Hybrid Variants Are Generally Strong

In nearly every model line (3, 5, 7 Series, i3, i8, etc.), Hybrid versions are consistently among the top-selling.

Suggests widespread consumer interest in fuel-efficient yet powerful cars.

🔸 3. Electric is Competitive, But Not Always Top

Electric variants (e.g., i3, i8) are strong in presence but do not top any category.

suggestion:

Range anxiety or infrastructure issues in some markets

Hybrid being a more transitional choice

🔸 4. Diesel and Petrol Still Hold Value

In models like M5, X5, and X6, Petrol and Diesel variants still show strong sales.

Strategic Takeaways:

📌 Hybrid tech investment should continue, especially in premium models like the 7 Series.

⚡ Electric versions should be promoted in regions with better infrastructure.

🛠️ Keep Diesel and Petrol options available for performance/SUV lines until demand clearly shifts.


**Regional Performance Analysis**

Filter: Cars with Price_USD > 50,000

Aggregate: Total Sales_Volume by Region

In [49]:
# Regional sales for high-priced cars
df[df['Price_USD'] > 50000].groupby('Region')['Sales_Volume'].sum().reset_index()

Region  Sales_Volume
0         Africa      32126173
1           Asia      33675286
2         Europe      32831246
3    Middle East      32837792
4  North America      32955309
5  South America      32254872

Insight: Highest Regional sales for high-priced cars in **Asia**

🌏 1. Asia Leads in Sales of High-End BMWs

Asia shows the highest total sales volume for cars priced above $50,000.

Indicates strong luxury car demand in countries like China, India, Japan, and South Korea.

Rising upper-middle class and preference for premium brands in emerging Asian markets could explain this trend.

🌍 2. Other Regions Are Close

North America, Middle East, and Europe have comparable sales volumes, each only marginally behind Asia.

Suggests that the global demand for luxury BMWs is widespread, not regionally isolated.

🌐 3. Africa and South America Lag Slightly

While volumes are slightly lower, the difference isn’t extreme.

May reflect developing markets with growing but still modest demand for high-end vehicles.

Potential for future growth as incomes rise and premium car infrastructure expands.

**transmission type demand:**
Highest demand cars: Mannual Transmission

In [67]:
df[df['Year'] >= 2020].groupby('Transmission')['Sales_Volume'].sum().reset_index()

Transmission  Sales_Volume
0    Automatic      41845997
1       Manual      43066966

In [68]:
df[df['Year'] < 2020].groupby('Transmission')['Sales_Volume'].sum().reset_index()

Transmission  Sales_Volume
0    Automatic      84167660
1       Manual      84295111

In [ ]:
**Pricing**

Filter: Cars with Mileage_KM < 50,000

Aggregate: Average Price_USD by Model

In [69]:
df[df['Mileage_KM'] < 50000].groupby('Model')['Price_USD'].mean().reset_index()


Model     Price_USD
0   3 Series  74888.180188
1   5 Series  74829.980427
2   7 Series  76368.543672
3         I3  75154.445643
4         I8  76480.128117
5         M3  74158.779661
6         M5  74760.368959
7         X1  74792.839789
8         X3  74724.899819
9         X5  74934.123144
10        X6  74468.269549

In [70]:
df[df['Mileage_KM'] >100000].groupby('Model')['Price_USD'].mean().reset_index()


Model     Price_USD
0   3 Series  75114.161687
1   5 Series  75616.961653
2   7 Series  75117.168443
3         I3  74243.349023
4         I8  74438.233846
5         M3  75286.215713
6         M5  74538.139966
7         X1  75380.385881
8         X3  75041.293578
9         X5  74758.470199
10        X6  74488.512029

Insights : 

🔍 1. Higher Prices for Low Mileage
Low-mileage cars generally command higher average prices than their high-mileage counterparts.

This reflects the market perception that less-used cars are more valuable, reliable, and likely to last longer.

🚘 2. Premium Models Stand Out
Models like i8, 7 Series, and X6 often have the highest average prices among low-mileage cars.

This is because these are luxury/performance lines, and buyers pay a premium for newer or less-used variants.

⚖️ 3. Smaller Models Still Hold Value
Even models like the 3 Series and X1, which are more entry-level, show strong resale value when mileage is low.

Suggests that these models are in demand among cost-conscious buyers seeking lightly used vehicles.

📊 4. Strategic Use in Pricing
Dealers or individual sellers can leverage mileage-based pricing to justify higher listings.

Buyers can use this info to negotiate better deals on higher-mileage versions of the same model.

In [83]:
df[(df['Year'] >2010) & (df['Engine_Size_L'] > 3.0)].groupby('Transmission')['Mileage_KM'].agg(['count', 'mean']) .reset_index()

Transmission  count           mean
0    Automatic  12813  100851.192383
1       Manual  13141   99107.045583

In [80]:
df[(df['Year'] < 2020) & (df['Engine_Size_L'] > 3.0)].groupby('Transmission')['Mileage_KM']  .agg(['count', 'mean']).reset_index()

Transmission  count           mean
0    Automatic   9145  100949.124221
1       Manual   9359   98770.759376

💡 Key Insights:

🔹 1. Manual Still Edges Out in Recent Models

Despite industry trends, manual big-engine BMWs post-2020,before 2020 slightly outnumber automatics.

Suggests a continued niche demand for manual transmissions among enthusiasts, especially for performance-oriented models.

🔹 2. Automatics Are Driven Slightly More
